aten/src/ATen/core/ivalue_inl.h

```diff
@@ -612,10 +625,18 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
       std::ostream& out,
       const Future& v);
 
+  void setElementType(TypePtr type) {
+    type_ = type;
+  }
+
   TypePtr elementType() const {
     return type_;
   }
 
+  void setDevices(std::vector<c10::Device> devices) {
+    devices_ = sortAndDeduplicateDevices(impl_, std::move(devices));
+  }
+
```

Comment:
I am **very** concerned by these two methods, for many reasons:
- I don't think they should be necessary (they weren't until now, and we've been using Future in a lot of scenarios already)
- They mean that the information queried from the future cannot be considered reliable. For example, we have places where we check that the type of a future is a certain thing, and then later we use it in a certain way. If we allow the type to change in between then our check becomes meaningless. (In a sense, this is a [TOCTOU race](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use)).
- The code, as it is now, is buggy and can lead to race conditions: Future is a type that's used when multiple threads are at play, and there's currently nothing preventing one thread to access the type of the devices while another thread modifies them. This could lead to garbage data, segfaults, and all the like.
- It could also be that this might lead the C++ compiler to do incorrect optimizations (though I'm not 100% sure). Just below we said that accessing type and devices is a const-qualified method, which the compiler could use to assume that after reading the value once it can be cached and reused, which is however _not_ the case if we allow the value to change underneath.

Reply:
Let me explain my decision:
I agree, the reason I've added them is that future is handled and set up by PG::Work class, subclasses can only modify the future itself, but not the pointer to it (as you've suggested to make future_ pointer const). 
However, in some cases, subclasses need to set some properties of future, e.g. GLOO needs to set a different return value type and NCCL would need to set which devices are involved in the future.

Alternative:
I share your concern here: we could have a data structure to gather all pieces necessary to create Future and create it lazily on the first access to getFuture().

Reply:
@lw  Curious, would the concern be mitigated if we ensured setElementType can only be called once and throw if not? This seems like it would eliminate the TOCTOU bug although I'm wondering what you think.

Reply:
I still think it would be a problem: the element type is a public property, and there are places where we retrieve and consume it, hence if we allow to delay setting it then every consumer needs to deal with the Future "being incomplete" and... what would one do in such a case?

If there were a strong reason for this use case we could find a way, but for the purposes of this PR I'm convinced we're always able to figure out the type at the time of the Future's creation, hence let's just do that?